include/deal.II/base/symmetric_tensor.h

```diff
@@ -1073,19 +1069,15 @@ inline SymmetricTensor<rank_, dim, Number>::SymmetricTensor(
 
 template <int rank_, int dim, typename Number>
 template <typename OtherNumber>
-inline SymmetricTensor<rank_, dim, Number>::SymmetricTensor(
+constexpr SymmetricTensor<rank_, dim, Number>::SymmetricTensor(
   const SymmetricTensor<rank_, dim, OtherNumber> &initializer)
-{
-  for (unsigned int i = 0; i < base_tensor_type::dimension; ++i)
-    data[i] =
-      internal::NumberType<typename base_tensor_type::value_type>::value(
-        initializer.data[i]);
```

Comment:
This is complicated on purpose, see https://github.com/dealii/dealii/commit/c29f87e2305. We need to be careful here.

Reply:
Looks like we need a test that checks whether `SymmetricTensor`s of various number types can be can be converted to each other.

By using the constructor that I defined
```c++
constexpr SymmetricTensor<rank_, dim, Number>::SymmetricTensor(
  const SymmetricTensor<rank_, dim, OtherNumber> &initializer)
  : data(initializer.data)
{}
```
which in turn calls the chain of converting constructors
```c++
Tensor<rank_, dim, Number>::Tensor(
  const Tensor<rank_, dim, OtherNumber> &initializer)
  : Tensor(initializer, std_cxx14::make_index_sequence<dim>{})
{}
```
and 
```c++
Tensor<0, dim, Number>::Tensor(const Tensor<0, dim, OtherNumber> &p)
  : Tensor{p.value}
{}
```
and 
```c++
Tensor<0, dim, Number>::Tensor(const OtherNumber &initializer)
  : value(internal::NumberType<Number>::value(initializer))
{}
```
 we shouldn't have any conversion issues.

Reply:
> Looks like we need a test that checks whether SymmetricTensors of various number types can be can be converted to each other.

I can create the test and submit it as a separate PR before we merge this one.